lock-analyzer/src/main.rs

```diff
@@ -0,0 +1,215 @@
+//! Analyzer for data produced by `wgpu-core`'s `observe_locks` feature.
+//!
+//! When `wgpu-core`'s `observe_locks` feature is enabled, if the
+//! `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is set to the
+//! path of an existing directory, then every thread that acquires a
+//! lock in `wgpu-core` will write its own log file to that directory.
+//! You can then run this program to read those files and summarize
+//! the results.
+//!
+//! This program also consults the `WGPU_CORE_LOCK_OBSERVE_DIR`
+//! environment variable to find the log files written by `wgpu-core`.
+//!
+//! See `wgpu_core/src/lock/observing.rs` for a general explanation of
+//! this analysis.
+
+use std::sync::Arc;
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet, HashMap},
+    fmt,
+    path::PathBuf,
+};
+
+use anyhow::{Context, Result};
+
+fn main() -> Result<()> {
+    let mut ranks: BTreeMap<u32, Rank> = BTreeMap::default();
+
+    let Ok(dir) = std::env::var("WGPU_CORE_LOCK_OBSERVE_DIR") else {
+        eprintln!(
+            "Please set the `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable\n\
+                   to the path of the directory containing the files written by\n\
+                   `wgpu-core`'s `observe_locks` feature."
+        );
+        anyhow::bail!("`WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is not set");
+    };
+    for entry in std::fs::read_dir(dir)? {
+        let entry = entry?;
+        let name = PathBuf::from(&entry.file_name());
+        let Some(extension) = name.extension() else {
+            eprintln!("Ignoring {}", name.display());
+            continue;
+        };
+        if extension != "ron" {
+            eprintln!("Ignoring {}", name.display());
+            continue;
+        }
+
+        let contents = std::fs::read(entry.path())
+            .with_context(|| format!("reading lock observations from {}", name.display()))?;
```

Comment:
nitpick: I'm puzzled by this context string pattern. This is what will be reported as the top-level error in the error chain, but it doesn't actually describe a problem that happened.

```suggestion
            .with_context(|| format!("failed to read lock observations from {}", name.display()))?;
```

TODO: Try this out, maybe the UX doesn't end up being what I think it will.